table/block.cc

```diff
@@ -505,9 +532,81 @@ bool IndexBlockIter::PrefixSeek(const Slice& target, uint32_t* index) {
   }
 }
 
+// NOTE: in hash seek, if the key is not found in the restart intervals,
+// the iterator will simply be set as "invalid", rather than returning
+// the key that is just pass the target key.
+void DataBlockIter::HashSeek(const Slice& target) {
+  assert(data_block_hash_index_);
+  Slice user_key = ExtractUserKey(target);
+  uint8_t entry = data_block_hash_index_->Seek(user_key);
+
+  if (entry == kNoEntry) {
+    current_ = restarts_;  // not found, Invalidate the iterator
+    return;
+  }
+
+  if (entry == kCollision) { // HashSeek not effective
+    status_ = Status::NotSupported("Hash Collision");
+    return;
+  }
+
+  uint32_t restart_index = entry;
+
+  // check if the key is in the restart_interval
+  SeekToRestartPoint(restart_index);
+  while (true) {
+    // Here we only linear seek the target key inside the restart interval.
+    // If a key does not exist inside a restart interval, we avoid
+    // further searching the block content accross restart interval boundary.
+    //
+    // TODO(fwu): check the left and write boundary of the restart interval
+    // to avoid linear seek a target key that is out of range.
+    //
+    // If using hash, we only linear search within the restart inteval.
+    if (!ParseNextDataKey(true /*within_restart_interval*/) ||
+        Compare(key_, target) >= 0) {
+      // we stop at the first potential matching user key.
+      break;
+    }
+  }
+
+  if ((current_ != restarts_) /* valid */ &&
+      // If the user key portion match do we consider key_ matches
+      user_comparator_->Compare(key_.GetUserKey(), user_key) == 0) {
+
+    // Here we are conservative and only support a limited set of cases
+    ValueType value_type = ExtractValueType(key_.GetKey());
+    if (value_type != ValueType::kTypeValue &&
+        value_type != ValueType::kTypeDeletion) {
+      status_ = Status::NotSupported("record type not supported");
+      return; // found, but not supported.
+    }
+
+    // Currently we do not fully support searching a key at specify snapshot.
+    // HashSeek only examine the first matched user_key. If the seqno is
+    // higher than the targe snapshot seqno, we just fall back to BinarySeek,
+    // without search further for the correct seqno.
+    uint64_t target_seqno = GetInternalKeySeqno(target);
+    uint64_t seqno = GetInternalKeySeqno(key_.GetKey());
+    if (target_seqno < seqno) {
+      status_ = Status::NotSupported("snapshot not fully supported");
+      // found, but not supported
+    }
+    return; // found
+  }
+  current_ = restarts_;  // not found, Invalidate the iterator
```

Comment:
There is difference between:
1. user key not found, or user key found but seqno all larger than seek key but it is not the end of the block.
2. user key found, all seqno all larger than seek key, but it is the end of the block.
In 1, Get() needs to stop. In 2, Get() needs to go to the next block. Please distinguish these two and handle accordingly.